design/deployment-strategies.md

```diff
@@ -0,0 +1,212 @@
+ ## Deployment strategies
+
+One of the key advantages of continuous delivery of application updates is the ability to quickly push updates into production for specific microservices, giving developers the ability to respond rapidly to changing business requirements. 
+
+With Azure DevOps, environments are first-class constructs, handling the underlying orchestration such as **deployment strategies**, verifying health checks, disabling old version and rolling out new version, traceability down artifact versions are key to continuous value delivery. 
+
+This proposal contains a number of interconnected elements. For example, an [environment](environment.md), a [deployment](deployment.md) job, primitive building blocks, and simplified YAML syntax to enable **safe deployments** in the majority of cases. Our approach to building a deployment orchestration language for **safe deployments** will be to get the fundamental building blocks right. 
+
+**runOnce**: default strategy, if not defined. 
+
+```yaml
+jobs:
+- deployment:
+  environment: musicCarnivalProd
+  pool:
+    name: musicCarnivalProdPool  
+  strategy:                 
+    runOnce:              
+      deploy:    
+        steps:             
+        - script: echo deploy web app...   
+ ```
+
+When deploying application updates it is important that the technique used to deliver update enables initialization, deploying the update, testing the updated version after routing traffic and in case of failure, run steps to restore to last known good version. We achieve this using fundamental building blocks viz., hooks where you can run your steps during deployment lifecycle. Following are descriptions of the lifecycle events where you can run a hook during a deployment lifecycle
+
+**preDeploy** – Use to run tasks before the deploy step is executed. A rollback is not possible at this point.
```

Comment:
This statement needs to be corrected. We need to run onFailure even for failures in preDeploy step.